### ISSUE TITLE
Dealing with Output Name Collisions

### ISSUE THREAD

**Issue by iandol** [htdebeer/pandocomatic#49]
Hi, this is quite minor, and could be handled by a cleanup script per template, but because we have multiple ways to generate the same output type  (e.g.generate a PDF via HTML and LaTeX), a later pandocomatic template output overwrites a previous template output. I was wondering, because pandocomatic know that the output of each template will result in the same filename, whether it could rename the outputs automatically. So say we have a file called test.md with `use-template: [html-pdf,latex-pdf]` — each template generates a PDF called test.pdf, so the latter overwrites the former. So it would be nice if the second template instead or output=test.pdf becomes test2.pdf or some such thing?

**Comment by htdebeer** [comment]
That could be possible, although it would be quite a specific piece of functionality. It would be nice if it could be made a bit more general.

Other options:

-   Another option could be to adapt the `output` option to allow "name templates". Something like:

    ```
    pandoc:
       from: markdown
       to: html
       output: "${base-name}2.pdf"
    ```
    As a disadvantage, you need to know beforehand the "name templates" you want to use in a specific situation.
-   Or use the "to" field as a "sub extension", so the example above (without the ``output`` option) would result in the filename `my_document.html.pdf`. This does not help if you generate two or more different "html.pdf" files from the same input file, of course.
As a disadvantage, you'd need to know these names in

**Comment by iandol** [comment]
The problem with specifying it in the template is as you say, we don't know if we will use it alone or with another template; this hardwiring is not flexible. And I can imagine generating two PDFs from the same source (with different LaTeX templates or HTML styles etc.). But this also has the same problem, if we only use one template then we want the default name, it is only when there is a collision that the name becomes important. These solution are only marginally better than using a cleanup script, so I imagine not worth the effort to implement.

**Comment by htdebeer** [comment]
I agree. It might be something to keep in mind when developing the next big release for pandocomatic, though. But before then we need to collect real-world use cases for pandocomatic in the wild to inventory common patterns to automate further.

**Comment by herculosh** [comment]
Hi @htdebeer 
I'll ask what the status is on this. 
I have exactly the same problem, I would like to create two pdf from one markdown, just passing different option to the template.  Now I ask myself how to do this best. 

I would prefer it, if I could enter the value in the external template in same the form `${base-name}2.pdf` as suggested by @iandol.

Thanks in Advance.

**Comment by htdebeer** [comment]
The status has not changed.

Ultimately, I do not yet know how to solve this problem in a general way that makes sense both for external pandocomatic templates and internal pandocomatic templates. Suggestions and ideas are welcome, though :-)

Currently I do think that to solve this issue I have to generalize and extend the workflow processing and configuration, allowing more fine-grained control about output, destinations, and other write actions. But that would make from pandocomatic an another beast all together.

**Comment by herculosh** [comment]
I could imagine that when creating the `outpout` files for Pandoc a simple string replacement with the file name could have a miraculous effect. 

Probably the modification placed in convert_file_command line [135](https://github.com/htdebeer/pandocomatic/blob/fb9e36ae946dc6f51ffbdd7692184397fdc6e3af/lib/pandocomatic/command/convert_file_command.rb#L135) would add the feature. 

Something like: 
```
@dst = @metadata.pandoc_options['output'] % {basename: File.basename(@dst, ".*")}
```

In the template we could define the output as follows.

```
output: "%{basename}-crazyfile.pdf"
```

I don't know Ruby that well and there's probably a logical reason why you don't choose this approach.

**Comment by htdebeer** [comment]
Technically it is not too hard to build something like this. The biggest problem I have is that this feature could impair the generality of templates. On the other hand, one could argue it is the responsibility of the user to know and understand her templates she is using, but that is not something to be careful with.

For example, I can understand having this feature to create two version of a PDF, one for USA audiences, and one for the rest of the world. Say I make two templates, `A4pdf` and `letterpdf`, each specifying the output as, respectively, `output: "${basename}-A4.pdf"` and `output: "${basename}-letter.pdf". And I use both templates in my file `leaflet.md`, like `use-template: ["A4pdf", "letterpdf"]`, I'd get `leaflet-A4.pdf` and `leaflet-letter.pdf`.

Similar use cases are using both small or large fonts, or having a file with solutions to exercises and one without, and so on. 

However, what should pandocomatic do when I override the `output` property in `leaflet.md`, say 

```yaml
use-template: ["A4pdf", "letterpdf"]
pandoc:
  output: "leaflet-final.pdf"
```

Would it make most sense to use:

- `leaflet-final.pdf`, the second template overriding the output of the first?
- `leaflet-final-A4.pdf` and `leaflet-final-letter.pdf`, interpreting `${basename}` differently depending on the templates and input file names?

Similarly, when, in the above example, the output is set to `"${basename}-final"`, should pandocomatic somehow combine the various "name templates"? (How?)

In either case, the biggest issue I see is that a user cannot infer from the template names and configurations of similar files what would happen. She does need to know and understand all templates involved as well as their cascading inheritance of properties.

Before we can implement this feature, we need to get straight how it works in various situations and we need to think of the usability of this feature. 

On top of that, if "name templates" with "${basename}" are implemented, it is possible similar features are requested as well. For example, include header files or CSS files using a glob pattern, i.e. `css: "assets/css/*.css"`. It would be nice to have a solution that can be appropriated to other situations. Ideas and suggestions for similar features are also welcome.
